sdk/src/main/scala/com/horizen/forge/Forger.scala

```diff
@@ -33,25 +37,68 @@ class Forger(settings: SidechainSettings,
              mainchainSynchronizer: MainchainSynchronizer,
              companion: SidechainTransactionsCompanion,
-             params: NetworkParams) extends Actor with ScorexLogging {
-
-  import Forger.ReceivableMessages._
-  import Forger._
-  import scorex.core.NodeViewHolder.ReceivableMessages.LocallyGeneratedModifier
+             val params: NetworkParams) extends Actor with ScorexLogging with TimeToEpochSlotConverter {
 
   val timeoutDuration: FiniteDuration = settings.scorexSettings.restApi.timeout
   implicit val timeout: Timeout = Timeout(timeoutDuration)
 
-  protected def getNodeView: View = {
-    val future = sidechainNodeViewHolderRef ? getCurrentNodeView
-    Await.result(future, timeoutDuration).asInstanceOf[View]
+  override def receive: Receive = {
+    processTryForgeNextBlockForEpochAndSlotMessage orElse {
+      case message: Any => log.error(s"Forger received strange message: ${message} from ${sender().path.name}")
+    }
+  }
+
+  protected def processTryForgeNextBlockForEpochAndSlotMessage: Receive = {
+    case TryForgeNextBlockForEpochAndSlot(consensusEpochNumber, consensusSlotNumber) => {
+      val forgingFunctionForEpochAndSlot: View => ForgeResult = tryToForgeNextBlock(consensusEpochNumber, consensusSlotNumber)
+
+      val forgeMessage =
+        GetDataFromCurrentView[SidechainHistory, SidechainState, SidechainWallet, SidechainMemoryPool, ForgeResult](forgingFunctionForEpochAndSlot)
+
+      val forgeMessageSender: ActorRef = sender()
+
+      val forgedBlockAsFuture = sidechainNodeViewHolderRef ? forgeMessage
+      forgedBlockAsFuture.onComplete{
+        case Success(forgeResult) => forgeMessageSender ! forgeResult //caller is responsible for applying new created block if required
+        case Failure(exception) => forgeMessageSender ! ForgeFailed(exception) //got some error during future completing
+      }
+    }
   }
 
-  protected def getNextBlockForgingInfo: Try[ForgingInfo] = Try {
-    val view = getNodeView
-    val parentId: Block.BlockId = view.history.bestBlockId
-    val timestamp: Long = Instant.now.getEpochSecond
+  protected def tryToForgeNextBlock(nextConsensusEpochNumber: ConsensusEpochNumber, nextConsensusSlotNumber: ConsensusSlotNumber)(view: View): ForgeResult = {
+    log.info(s"Try to forge block for epoch ${nextConsensusEpochNumber} with slot ${nextConsensusSlotNumber}")
+    val bestBlockId = view.history.bestBlockId
+    val bestBlockInfo = view.history.bestBlockInfo
+    val bestBlockEpochAndSlot = timestampToEpochAndSlot(bestBlockInfo.timestamp)
 
+    val nextBockTimestamp = getTimeStampForEpochAndSlot(nextConsensusEpochNumber, nextConsensusSlotNumber)
+    val nextBlockEpochAndSlot = EpochAndSlot(nextConsensusEpochNumber, nextConsensusSlotNumber)
+    if(bestBlockEpochAndSlot > nextBlockEpochAndSlot) {
+      log.warn(s"Try to forge block with epochAndSlot ${nextBlockEpochAndSlot} but current best block epochAndSlot are: ${bestBlockEpochAndSlot}")
```

Comment:
`bestBlockEpochAndSlot > nextBlockEpochAndSlot` -> `bestBlockEpochAndSlot >= nextBlockEpochAndSlot`
Implement `>=` and `<=` operations in `EpochAndSlot`.
Here we should return an error, no sense to generate a block, that will fail timestamp check in `ConsensusValidator`.